File: C++/kindle-oo-design.cpp
/*
Design Kindle, which can support 3 type of book format: PDF, MOBI and EPUB.

- Consider using ENUM for book format.
- Consider using simple factory to create reader for each format.

Example
Input:
readBook("EPUB")
readBook("PDF")
readBook("MOBI")

Output:
Using EPUB reader, book content is: epub
Using PDF reader, book content is: pdf
Using MOBI reader, book content is: mobi

Solution:
    EBookReaderFactory *factory = new EBookReaderFactory;
    EBookReader *ebook = factory->createReader(Book *b); // switch to get the reader subclass
    ebook->readbook();
*/

const char* names[] = { "epub","pdf","mobi" };

enum Format { EPUB, PDF, MOBI };

class Book {
private:
    Format format;

public:
    Book(Format format) {
        this->format = format;
    }

    Format getFormat() {
        return format;
    }
};

class EBookReader {
protected:
    Book *book;

public:

    EBookReader(Book *b) {
        this->book = b;
    }

    string readBook() {
        return reader->displayReaderType() + ", book content is: " + names[book->getFormat()];
    }

    virtual string displayReaderType() = 0;
};

class EpubReader :public EBookReader {
public:
    EpubReader(Book *b):EBookReader(b){}

    string displayReaderType() {
        return "Using EPUB reader";
    }
};

class MobiReader :public EBookReader {
public:
    MobiReader(Book *b):EBookReader(b){}

    string displayReaderType() {
        return "Using MOBI reader";
    }
};

class PdfReader :public EBookReader {
public:
    PdfReader(Book *b):EBookReader(b){}

    string displayReaderType() {
        return "Using PDF reader";
    }
};

class EBookReaderFactory {
public:

    EBookReader *createReader(Book *b) {
        switch (b->getFormat()) {
            case EPUB:
                return new EpubReader(b);
            case MOBI:
                return new MobiReader(b);
            case PDF:
                return new PdfReader(b);
            default:
                return NULL;
        }
    }
};

class Kindle {
private:
    vector<Book *> *books;
    EBookReaderFactory *readerFactory;

public:
    Kindle() {
        books = new vector<Book *>;
        readerFactory = new EBookReaderFactory;
    }

    string readBook(Book *book) {
        EBookReader *reader = readerFactory->createReader(book);
        return reader->readBook();
    }

    void downloadBook(Book *b) {
        books->push_back(b);
    }

    void uploadBook(Book *b) {
        books->push_back(b);
    }

    void deleteBook(Book *b) {
        vector<Book *>::iterator it;
        for (it = books->begin(); it != books->end(); it++) {
            if ((*it)->getFormat() == b->getFormat()) {
                books->erase(it);
                return;
            }
        }
    }
};

File: Java/kindle-oo-design.java
import java.util.ArrayList;
import java.util.List;

public class Kindle {
    private List<Book> books;
    private EBookReaderFactory readerFactory;

    public Kindle() {
        books = new ArrayList<>();
        readerFactory = new EBookReaderFactory();
    }

    public String readBook(Book book) throws Exception {
        EBookReader reader = readerFactory.createReader(book);
        if (reader == null) {
            throw new Exception("Can't read this format");
        }
        return reader.readBook();
    }

    public void downloadBook(Book b) {
        books.add(b);
    }

    public void uploadBook(Book b) {
        books.add(b);
    }

    public void deleteBook(Book b) {
        books.remove(b);
    }
}

enum Format {
    EPUB("epub"), PDF("pdf"), MOBI("mobi");

    private String content;

    Format(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }
}

class Book {
    private Format format;

    public Book(Format format) {
        this.format = format;
    }

    public Format getFormat() {
        return format;
    }
}

class EBookReaderFactory {
    public EBookReader createReader(Book b) {
        switch (b.getFormat()) {
            case EPUB:
                return new EpubReader(b);
            case MOBI:
                return new MobiReader(b);
            case PDF:
                return new PdfReader(b);
            default:
                return null;
        }
    }
}

abstract class EBookReader {
    protected Book book;
    
    public EBookReader(Book b){
        this.book = b;
    }
    
    public String readBook() {
        return displayReaderType() + ", book content is: " + this.book.getFormat().getContent();
    }

    public abstract String displayReaderType();
}

class EpubReader extends EBookReader{
    public EpubReader(Book b) {
        super(b);
    }

    @Override
    public String displayReaderType() {
        return "Using EPUB reader";
    }
}

class MobiReader extends EBookReader {
    public MobiReader(Book b) {
        super(b);
    }

    @Override
    public String displayReaderType() {
        return "Using MOBI reader";
    }
}

class PdfReader extends EBookReader{
    public PdfReader(Book b) {
        super(b);
    }

    @Override
    public String displayReaderType() {
        return "Using PDF reader";
    }
}
